identity/v3userpass.go

```diff
@@ -0,0 +1,182 @@
+package identity
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	goosehttp "gopkg.in/goose.v1/http"
+)
+
+// v3AuthWrapper wraps the v3AuthRequest to perform v3 authentication.
+type v3AuthWrapper struct {
+	Auth v3AuthRequest `json:"auth"`
+}
+
+// v3AuthRequest contains the authentication request.
+type v3AuthRequest struct {
+	Identity v3AuthIdentity `json:"identity"`
+	Scope    *v3AuthScope   `json:"scope,omitempty"`
+}
+
+// v3AuthIdentity contains the identity portion of an authentication
+// request.
+type v3AuthIdentity struct {
+	Methods  []string        `json:"methods"`
+	Password *v3AuthPassword `json:"password,omitempty"`
+	Token    *v3AuthToken    `json:"token,omitempty"`
+}
+
+// v3AuthPassword contains a password authentication request.
+type v3AuthPassword struct {
+	User v3AuthUser `json:"user"`
+}
+
+// v3AuthUser contains the user part of a password authentication
+// request..
+type v3AuthUser struct {
+	Domain   *v3AuthDomain `json:"domain,omitempty"`
+	ID       string        `json:"id,omitempty"`
+	Name     string        `json:"name,omitempty"`
+	Password string        `json:"password"`
+}
+
+// v3AuthDomain contains a domain definition of an authentication
+// request.
+type v3AuthDomain struct {
+	ID   string `json:"id,omitempty"`
+	Name string `json:"name,omitempty"`
+}
+
+// v3AuthToken contains the token to use for token authentication.
```

Comment:
I'm not familiar with the details, but does this cover all types of tokens? the uuid, pki and fernet one?